galley/pkg/config/analysis/analyzers/auth/mtls.go

```diff
@@ -0,0 +1,355 @@
+//
+// The analyzer tries to act more generally by imagining service-to-service
+// traffic and detecting whether or not there's a conflict with regards to mTLS
+// policy. This means it will also detect explicit misconfigurations as well.
+//
+// Note this is very similar to `istioctl authn tls-check`; however this
+// inspection is all done via analyzing configuration rather than requiring a
+// connection to pilot.
+type MTLSAnalyzer struct{}
+
+// Compile-time check that this Analyzer correctly implements the interface
+var _ analysis.Analyzer = &MTLSAnalyzer{}
+
+// Metadata implements Analyzer
+func (s *MTLSAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name: "auth.MTLSAnalyzer",
+
+		// Each analyzer should register the collections that it needs to use as input.
+		Inputs: collection.Names{
+			metadata.K8SCoreV1Pods,
+			metadata.K8SCoreV1Namespaces,
+			metadata.K8SCoreV1Services,
+			metadata.IstioAuthenticationV1Alpha1Meshpolicies,
+			metadata.IstioAuthenticationV1Alpha1Policies,
+			metadata.IstioMeshV1Alpha1MeshConfig,
+			metadata.IstioNetworkingV1Alpha3Destinationrules,
+		},
+	}
+}
+
+// Analyze implements Analyzer
+func (s *MTLSAnalyzer) Analyze(c analysis.Context) {
+	// TODO Reuse pilot logic as a library rather than reproducing its logic
+	// here.
+
+	// If autoMTLS is turned on, bail out early as the logic used below does not
+	// reason about its usage.
+	autoMtlsEnabled := false
+	rootNamespace := "istio-system"
+	c.ForEach(metadata.IstioMeshV1Alpha1MeshConfig, func(r *resource.Entry) bool {
+		mc := r.Item.(*meshconfig.MeshConfig)
+		if mc.GetEnableAutoMtls() != nil && mc.GetEnableAutoMtls().Value {
+			autoMtlsEnabled = true
+		}
+
+		if mc.GetRootNamespace() != "" {
+			rootNamespace = mc.GetRootNamespace()
+		}
+		return true
+	})
+
+	if autoMtlsEnabled {
+		return
+	}
+
+	// Loop over all services, building up a list of selectors for each. This is
+	// used to determine which pods are in which services, and determine whether
+	// or not the sidecar is fully enmeshed. If a service doesn't have a
+	// sidecar, then we always treat it as having an explicit "plaintext" policy
+	// regardless of the service/namespace/mesh policy.
+
+	var targetServices []mtls.TargetService
+	fqdnsWithoutSidecars := make(map[string]struct{})
+	// Keep track of each fqdn -> port name -> port number. This is because
+	// the Policy object lets you target a port name, but DR requires port
+	// number. Tracking this means we can normalize to port number later.
+	fqdnToNameToPort := make(map[string]map[string]uint32)
+
+	c.ForEach(metadata.K8SCoreV1Services, func(r *resource.Entry) bool {
+		svcNs, svcName := r.Metadata.Name.InterpretAsNamespaceAndName()
+
+		// Skip the istio control plane. It doesn't obey Policy/MeshPolicy MTLS
+		// rules in general and instead is controlled by the mesh option
+		// 'controlPlaneSecurityEnabled'.
+		if svcNs == "istio-system" {
```

Comment:
Yes. Currently not supported with the analyze command yet though (probably will be supplied via command-line argument).

Reply:
Can you get the control plane from the provided mesh config? That's what the rootNamespace field is for, right?

Reply:
rootNamespace appears to be different than control plane namespace:
```
The namespace to treat as the administrative root namespace for Istio configuration. When processing a leaf namespace Istio will search for declarations in that namespace first and if none are found it will search in the root namespace. Any matching declaration found in the root namespace is processed as if it were declared in the leaf namespace.

The precise semantics of this processing are documented on each resource type.
```

Reply:
You're right, it's not precisely the same thing. Drat. 

Istio namespace actually is already a CLI option included in all istioctl commands, and we use it already, so we know where to look for the mesh config on a live cluster. Making that information available to the analysis context is definitely doable, just takes more plumbing. (And yeah, it probably makes sense to have that as a follow-up).

Reply:
Concretely, for the record, what we'd probably want to do is:

* Add a field to `AnalyzingDistributorSettings` for istio system namespace. In the live galley case, plumb through the value of `DeploymentAndServiceNamespace`. For local analysis, plumb through istioNamespace. 
* Add a field representing the istio namespace directly to the analysis context object, and a getter method that returns it.